src/coord/tests/sql.rs

```diff
@@ -39,72 +42,89 @@ use mz_sql::DEFAULT_SCHEMA;
 async fn datadriven() {
     datadriven::walk_async("tests/testdata", |mut f| async {
         let data_dir = TempDir::new().unwrap();
-        let mut catalog = Catalog::open_debug(data_dir.path(), NOW_ZERO.clone())
-            .await
-            .unwrap();
-        f.run(|test_case| -> String {
-            match test_case.directive.as_str() {
-                "add-table" => {
-                    let id = catalog.allocate_user_id().unwrap();
-                    let oid = catalog.allocate_oid().unwrap();
-                    let database_id = catalog
-                        .resolve_database(DEFAULT_DATABASE_NAME)
-                        .unwrap()
-                        .id();
-                    let database_spec = ResolvedDatabaseSpecifier::Id(database_id);
-                    let schema_spec = catalog
-                        .resolve_schema_in_database(&database_spec, DEFAULT_SCHEMA, SYSTEM_CONN_ID)
-                        .unwrap()
-                        .id
-                        .clone();
-                    catalog
-                        .transact(
-                            vec![Op::CreateItem {
-                                id,
-                                oid,
-                                name: QualifiedObjectName {
-                                    qualifiers: ObjectQualifiers {
-                                        database_spec,
-                                        schema_spec,
+        // The datadriven API has FnMuts, and doesn't express to Rust that it will
+        // complete each run_async call before doing the next, so we have to wrap
+        // the catalog. Datadriven could be changed to maybe take some context that
+        // it passes as a &mut to each test_case invocation, act on a stream of
+        // test_cases, or take and return a Context.
```

Comment:
```suggestion
        // The datadriven API takes an `FnMut` closure, and can't express to Rust that it will
        // finish polling each returned future before calling the closure again, so we have to wrap
        // the catalog in a share-able type. Datadriven could be changed to maybe take some context that
        // it passes as a &mut to each test_case invocation, act on a stream of
        // test_cases, or take and return a Context. This is just a test, so the performance hit of this
        // doesn't matter (and in practice there will be no contention)
```